File: Problemset/1652-Defuse the Bomb/Solution.rb
# @param {Integer[]} code
# @param {Integer} k
# @return {Integer[]}
def decrypt(code, k)
  n = code.length
  s = k > 0 ? 1 : n + k
  e = k > 0 ? k + 1 : n
  sum = code[s...e].sum
  ret = Array.new(n)

  (0...n).each do |i|
    ret[i] = sum
    sum += code[(e + i) % n] - code[(s + i) % n]
  end

  ret
end

File: Problemset/1652-Defuse the Bomb/Solution.rs
impl Solution {
    pub fn decrypt(code: Vec<i32>, k: i32) -> Vec<i32> {
        let n = code.len();
        let s = if k > 0 { 1 } else { (n as i32 + k) as usize };
        let e = if k > 0 { k as usize + 1 } else { n };
        let mut sum = code[s..e].iter().sum();
        let mut ret = vec![0; n];

        for i in 0..n {
            ret[i] = sum;
            sum += code[(e + i) % n] - code[(s + i) % n];
        }

        ret
    }
}
